pkg/rkeworker/docker.go

```diff
@@ -127,87 +127,122 @@ func remove(ctx context.Context, c *client.Client, id string) error {
-		Env:         inspect.Config.Env,
-		Image:       inspect.Config.Image,
-		Binds:       inspect.HostConfig.Binds,
-		NetworkMode: string(inspect.HostConfig.NetworkMode),
-		PidMode:     string(inspect.HostConfig.PidMode),
-		Privileged:  inspect.HostConfig.Privileged,
-		VolumesFrom: inspect.HostConfig.VolumesFrom,
-		Labels:      inspect.Config.Labels,
+	actualProcess := v3.Process{
+		Command:     actualDockerInspect.Config.Entrypoint,
+		Args:        actualDockerInspect.Config.Cmd,
+		Env:         actualDockerInspect.Config.Env,
+		Image:       actualDockerInspect.Config.Image,
+		Binds:       actualDockerInspect.HostConfig.Binds,
+		NetworkMode: string(actualDockerInspect.HostConfig.NetworkMode),
+		PidMode:     string(actualDockerInspect.HostConfig.PidMode),
+		Privileged:  actualDockerInspect.HostConfig.Privileged,
+		VolumesFrom: actualDockerInspect.HostConfig.VolumesFrom,
+		Labels:      actualDockerInspect.Config.Labels,
 	}
 
-	if len(p.Command) == 0 {
-		p.Command = newProcess.Command
+	if len(expectedProcess.Command) == 0 {
+		expectedProcess.Command = actualProcess.Command
 	}
-	if len(p.Args) == 0 {
-		p.Args = newProcess.Args
+	if len(expectedProcess.Args) == 0 {
+		expectedProcess.Args = actualProcess.Args
 	}
-	if len(p.Env) == 0 {
-		p.Env = newProcess.Env
+	if len(expectedProcess.Env) == 0 {
+		expectedProcess.Env = actualProcess.Env
 	}
-	if p.NetworkMode == "" {
-		p.NetworkMode = newProcess.NetworkMode
+	if expectedProcess.NetworkMode == "" {
+		expectedProcess.NetworkMode = actualProcess.NetworkMode
 	}
-	if p.PidMode == "" {
-		p.PidMode = newProcess.PidMode
+	if expectedProcess.PidMode == "" {
+		expectedProcess.PidMode = actualProcess.PidMode
 	}
-	if len(p.Labels) == 0 {
-		p.Labels = newProcess.Labels
+	if len(expectedProcess.Labels) == 0 {
+		expectedProcess.Labels = actualProcess.Labels
 	}
 
 	// Don't detect changes on these fields
-	newProcess.Name = p.Name
-	newProcess.HealthCheck.URL = p.HealthCheck.URL
-	newProcess.RestartPolicy = p.RestartPolicy
+	actualProcess.Name = expectedProcess.Name
+	actualProcess.HealthCheck.URL = expectedProcess.HealthCheck.URL
+	actualProcess.RestartPolicy = expectedProcess.RestartPolicy
 
 	changed := false
-	t := reflect.TypeOf(newProcess)
+	t := reflect.TypeOf(actualProcess)
 	for i := 0; i < t.NumField(); i++ {
 		f := t.Field(i)
 		if f.Name == "Command" {
-			leftMap := sliceToMap(p.Command)
-			rightMap := sliceToMap(newProcess.Command)
+			leftMap := sliceToMap(expectedProcess.Command)
+			rightMap := sliceToMap(actualProcess.Command)
+
 			if reflect.DeepEqual(leftMap, rightMap) {
 				continue
 			}
 		} else if f.Name == "Env" {
-			// all env
-			allEnv := append(p.Env, imageInspect.Config.Env...)
-			leftMap := sliceToMap(allEnv)
-			rightMap := sliceToMap(newProcess.Env)
-			if reflect.DeepEqual(leftMap, rightMap) {
+			expectedEnvs := make(map[string]string, 8)
+			for _, env := range defaultDockerInspect.Config.Env {
+				es := strings.Split(env, "=")
```

Comment:
can you use strings.SplitN(env, "=", 2)? also i suggest doing a length check.

Reply:
OK